### PR TITLE
Enhance the scan status and error reporting from EMA

### PR DESCRIPTION
**Pull request by AHabes**
### What is the purpose of this change?

Enhance the scan status and error reporting from EMA

### How was this change implemented?

1. Refactoring the scan status route to make it simpler and separate the overall status from the per-route status.
2. Fix the out of order scan status and scan data messages by changing the event portal endpoint to become a direct component
3. Fix the complete scan status for EMA in offline mode (complete status is persisted in DB)
4. Fix the complete scan status for EMA in offline mode (complete status is logged into the console)
5. Implement the complete status for empty scan types and its descendent in the RouteBundle hierarchy after parsing the hierarchy (no depth constraint) and extract all possible root-to-leaf paths
6. Introduce a new exception handling route at global level.
7. Introduce custom exception classes for Scan status and scan data publisher.
8. Introduce exception handlers for scan status and scan data publisher
9. Improve the unit tests for Scan manager and scan service

### How was this change tested?

Manual + unit tests

### Is there anything the reviewers should focus on/be aware of?
N/A

### CODE REVIEW

File: service/application/src/main/java/com/solace/maas/ep/event/management/agent/processor/EmptyScanEntityProcessorImpl.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                    log.info("Scan request [{}]: Encountered an empty scan type [{}].", scanId, childScanType);
+                    sendCompleteStatusForScanType(scanId, messagingServiceId, childScanType);
+                    log.info("Scan request [{}]: The status of [{}]" + " is: [{}].", scanId, childScanType, ScanStatus.COMPLETE);
+                });
+            }
+        });
+    }
+
+    public void sendCompleteStatusForScanType(String scanId, String messagingServiceId, String scanType) {
+        producerTemplate.send("direct:processScanStatusAsComplete?block=false&failIfNoConsumers=false", exchange -> {
```
Comment (slunelsolace) [inline service/application/src/main/java/com/solace/maas/ep/event/management/agent/processor/EmptyScanEntityProcessorImpl.java:77]: We shouldn't be using Producer Template inside of Processors.

File: service/application/src/main/java/com/solace/maas/ep/event/management/agent/processor/ScanStatusPerRouteProcessor.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        topicDetails.put("messagingServiceId", messagingServiceId);
+        topicDetails.put("scanId", scanId);
+        topicDetails.put("status", status.name());
+        topicDetails.put("scanDataType", scanType);
+
+        ScanDataStatusMessage scanDataStatusMessage = new
+                ScanDataStatusMessage(orgId, scanId, status.name(), description, scanType);
+
+        try {
+            scanStatusPublisher.sendScanDataStatus(scanDataStatusMessage, topicDetails);
```
Comment (slunelsolace) [inline service/application/src/main/java/com/solace/maas/ep/event/management/agent/processor/ScanStatusPerRouteProcessor.java:62]: Why call this inside of the processor? Can't we use the route instead?

**Review by slunelsolace — CHANGES_REQUESTED** [review]
Can we remove calling ProducerTemplate from Processors? We should be using the Route to forward the body after the Processor sets it.

File: service/application/src/main/java/com/solace/maas/ep/event/management/agent/processor/EmptyScanEntityProcessorImpl.java (hunk 1 above)
Comment (AHabes) [reply service/application/src/main/java/com/solace/maas/ep/event/management/agent/processor/EmptyScanEntityProcessorImpl.java]: Thanks for the feedback. The processor and associated routes have been updated.

File: service/application/src/main/java/com/solace/maas/ep/event/management/agent/processor/ScanStatusPerRouteProcessor.java (hunk 2 above)
Comment (AHabes) [reply service/application/src/main/java/com/solace/maas/ep/event/management/agent/processor/ScanStatusPerRouteProcessor.java:62]: Thanks for the feedback. Actually, there are several "publishing processors" that are used to publish the scan data, scan status, scan logs, and import data. If we would like to replace all these processors, then I suggested to implement this change in a separate PR, as it is going to be a bigger task. 

Kindly let me know if this change is necessary in the near future, so I can keep track of it.

**Review by slunelsolace — APPROVED** [review]

**Pull request merged**